File: app/src/main/java/xzr/konabess/adapters/ChipsetSelectorAdapter.java
package xzr.konabess.adapters;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.chip.Chip;
import com.google.android.material.textview.MaterialTextView;

import java.util.List;

import xzr.konabess.KonaBessCore;
import xzr.konabess.R;

public class ChipsetSelectorAdapter extends RecyclerView.Adapter<ChipsetSelectorAdapter.ViewHolder> {

    public interface OnChipsetSelectedListener {
        void onChipsetSelected(KonaBessCore.Dtb dtb);
    }

    private final List<KonaBessCore.Dtb> dtbList;
    private final Activity activity;
    private final int recommendedIndex;
    private final Integer currentlySelectedId; // Can be null
    private final OnChipsetSelectedListener listener;

    // Constructor for initial device preparation (no currently selected)
    public ChipsetSelectorAdapter(List<KonaBessCore.Dtb> dtbList, Activity activity, 
                                  int recommendedIndex, OnChipsetSelectedListener listener) {
        this(dtbList, activity, recommendedIndex, null, listener);
    }

    // Constructor for chipset switching (with currently selected)
    public ChipsetSelectorAdapter(List<KonaBessCore.Dtb> dtbList, Activity activity, 
                                  int recommendedIndex, Integer currentlySelectedId,
                                  OnChipsetSelectedListener listener) {
        this.dtbList = dtbList;
        this.activity = activity;
        this.recommendedIndex = recommendedIndex;
        this.currentlySelectedId = currentlySelectedId;
        this.listener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_chipset_option, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        KonaBessCore.Dtb dtb = dtbList.get(position);
        
        // Set chipset name
        String chipsetName = dtb.id + " " + xzr.konabess.ChipInfo.name2chipdesc(dtb.type, activity);
        holder.chipsetName.setText(chipsetName);
        
        // Set subtitle
        holder.chipsetSubtitle.setText("DTB Index: " + dtb.id);
        
        // Show recommended badge if this is the recommended chipset
        if (dtb.id == recommendedIndex) {
            holder.recommendedBadge.setVisibility(View.VISIBLE);
        } else {
            holder.recommendedBadge.setVisibility(View.GONE);
        }
        
        // Show currently selected badge if this is the currently selected chipset
        if (currentlySelectedId != null && dtb.id == currentlySelectedId) {
            holder.selectedBadge.setVisibility(View.VISIBLE);
        } else {
            holder.selectedBadge.setVisibility(View.GONE);
        }
        
        // Set click listener
        holder.cardView.setOnClickListener(v -> {
            if (listener != null) {
                listener.onChipsetSelected(dtb);
            }
        });
        
        // Add subtle elevation on press
        holder.cardView.setOnTouchListener((v, event) -> {
            switch (event.getAction()) {
                case android.view.MotionEvent.ACTION_DOWN:
                    holder.cardView.setCardElevation(8f);
                    break;
                case android.view.MotionEvent.ACTION_UP:
                case android.view.MotionEvent.ACTION_CANCEL:
                    holder.cardView.setCardElevation(0f);
                    break;
            }
            return false;
        });
    }

    @Override
    public int getItemCount() {
        return dtbList.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        MaterialCardView cardView;
        ImageView chipsetIcon;
        MaterialTextView chipsetName;
        MaterialTextView chipsetSubtitle;
        Chip recommendedBadge;
        Chip selectedBadge;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            cardView = (MaterialCardView) itemView;
            chipsetIcon = itemView.findViewById(R.id.chipset_icon);
            chipsetName = itemView.findViewById(R.id.chipset_name);
            chipsetSubtitle = itemView.findViewById(R.id.chipset_subtitle);
            recommendedBadge = itemView.findViewById(R.id.recommended_badge);
            selectedBadge = itemView.findViewById(R.id.selected_badge);
        }
    }
}

File: app/src/main/java/com/ireddragonicy/konabessnext/ui/adapters/ChipsetSelectorAdapter.kt
package com.ireddragonicy.konabessnext.ui.adapters

import android.app.Activity
import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView
import com.google.android.material.chip.Chip
import com.google.android.material.textview.MaterialTextView
import com.google.android.material.color.MaterialColors
import com.ireddragonicy.konabessnext.R
import com.ireddragonicy.konabessnext.model.Dtb

class ChipsetSelectorAdapter(
    private val dtbList: List<Dtb>,
    private val activity: Activity,
    private val activeDtbId: Int, // Changed name for clarity, was recommendedIndex
    private val currentlySelectedId: Int?,
    private val listener: OnChipsetSelectedListener?
) : RecyclerView.Adapter<ChipsetSelectorAdapter.ViewHolder>() {

    constructor(
        dtbList: List<Dtb>,
        activity: Activity,
        activeDtbId: Int,
        listener: OnChipsetSelectedListener?
    ) : this(dtbList, activity, activeDtbId, null, listener)

    fun interface OnChipsetSelectedListener {
        fun onChipsetSelected(dtb: Dtb)
    }

    companion object {
        const val MANUAL_SETUP_ID = -999
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == dtbList.size) 1 else 0 // 1 for manual setup, 0 for item
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_chipset_option, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (getItemViewType(position) == 1) {
            // Manual Setup Item
            holder.chipsetName.text = "Manual Setup / Deep Scan"
            holder.chipsetSubtitle.text = "Configure unsupported device manually"
            holder.chipsetIcon.setImageResource(R.drawable.ic_search) 
            holder.recommendedBadge.visibility = View.GONE
            holder.selectedBadge.visibility = View.GONE
            
            holder.cardView.setOnClickListener {
                listener?.onChipsetSelected(Dtb(MANUAL_SETUP_ID, com.ireddragonicy.konabessnext.model.ChipDefinition("manual", "Manual", 0, false, 0, null, "", 0, mapOf()))) 
            }
        } else {
            val dtb = dtbList[position]

            // Set chipset name
            val chipsetName = "${dtb.id} ${dtb.type.name}"
            holder.chipsetName.text = chipsetName

            // Set subtitle
            holder.chipsetSubtitle.text = "DTB Index: ${dtb.id}"

            // Show ACTIVE badge (System running)
            if (dtb.id == activeDtbId) {
                holder.recommendedBadge.visibility = View.VISIBLE
                holder.recommendedBadge.text = "ACTIVE"
                holder.recommendedBadge.setChipIconResource(R.drawable.ic_check) // Or power icon
                // Style as Green/Success
                val colorGreen = activity.getColor(android.R.color.holo_green_dark)
                // We'd ideally use dynamic colors, but for distinct "Active", green is standard.
                // Or use TertiaryContainer from theme.
            } else {
                holder.recommendedBadge.visibility = View.GONE
            }

            // Show currently selected badge (Editing)
            if (currentlySelectedId != null && dtb.id == currentlySelectedId) {
                holder.selectedBadge.visibility = View.VISIBLE
            } else {
                holder.selectedBadge.visibility = View.GONE
            }

            // Set click listener
            holder.cardView.setOnClickListener {
                listener?.onChipsetSelected(dtb)
            }
        }

        // Add subtle elevation on press
        holder.cardView.setOnTouchListener { _, event ->
            when (event.action) {
                android.view.MotionEvent.ACTION_DOWN -> holder.cardView.cardElevation = 8f
                android.view.MotionEvent.ACTION_UP, android.view.MotionEvent.ACTION_CANCEL -> holder.cardView.cardElevation = 0f
            }
            false
        }
    }

    override fun getItemCount(): Int {
        return dtbList.size + 1
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cardView: MaterialCardView = itemView as MaterialCardView
        val chipsetIcon: ImageView = itemView.findViewById(R.id.chipset_icon)
        val chipsetName: MaterialTextView = itemView.findViewById(R.id.chipset_name)
        val chipsetSubtitle: MaterialTextView = itemView.findViewById(R.id.chipset_subtitle)
        val recommendedBadge: Chip = itemView.findViewById(R.id.recommended_badge)
        val selectedBadge: Chip = itemView.findViewById(R.id.selected_badge)
    }
}